Snowflake.Data/Core/Converter/JsonToStructuredTypeConverter.cs

```diff
@@ -0,0 +1,423 @@
+            {
+                while (jsonEnumerator.MoveNext())
+                {
+                    var jsonPropertyWithValue = jsonEnumerator.Current;
+                    var fieldValue = jsonPropertyWithValue.Value;
+                    var key = IsTextMetadata(keyMetadata) || IsFixedMetadata(keyMetadata)
+                        ? ConvertToUnstructuredType(keyMetadata, keyType, jsonPropertyWithValue.Key)
+                        : throw new Exception("Could not recognize key type for map");
+                    var value = ConvertToStructuredOrUnstructuredValue(valueType, fieldMetadata, fieldValue);
+                    result.Add(key, value);
+                }
+            }
+            return result;
+        }
+
+        private static object ConvertToStructuredOrUnstructuredValue(Type valueType, FieldMetadata fieldMetadata, JToken fieldValue)
+        {
+            try
+            {
+                if (IsObjectMetadata(fieldMetadata) && IsStructured(fieldMetadata))
+                {
+                    return ConvertToObject(valueType, fieldMetadata.fields, fieldValue);
+                }
+
+                if (IsArrayMetadata(fieldMetadata) && IsStructured(fieldMetadata))
+                {
+                    var nestedType = GetNestedType(valueType);
+                    return ConvertToArray(valueType, nestedType, fieldMetadata.fields, fieldValue);
+                }
+
+                if (IsMapMetadata(fieldMetadata) && IsStructured(fieldMetadata))
+                {
+                    var keyValueTypes = GetMapKeyValueTypes(valueType);
+                    return ConvertToMap(valueType, keyValueTypes[0], keyValueTypes[1], fieldMetadata.fields, fieldValue);
+                }
+
+                return ConvertToUnstructuredType(fieldMetadata, valueType, fieldValue);
+            }
+            catch (Exception e)
+            {
+                if (e is SnowflakeDbException)
+                    throw;
+                throw new SnowflakeDbException(SFError.STRUCTURED_TYPE_READ_ERROR, $"when reading '{fieldMetadata.name}' property", e.Message);
+            }
+        }
+
+        private static Type GetNestedType(Type type)
+        {
+            if (type.IsArray)
+            {
+                return type.GetElementType();
+            }
+            if (IsListType(type) || IsIListType(type))
+            {
+                return type.GenericTypeArguments[0];
+            }
+            throw new Exception("neither array nor list");
+        }
+
+        private static Type[] GetMapKeyValueTypes(Type type)
+        {
+            var genericParamWithTwoArguments = type.IsGenericType && type.GenericTypeArguments.Length == 2;
+            if (!genericParamWithTwoArguments)
+            {
+                throw new Exception("Could not get key and value types");
+            }
+            return type.GenericTypeArguments;
+        }
+
+        private static Type MakeArrayType(Type type, Type elementType)
```

Comment:
Maybe GetArrayType to have consistent names for such methods (see above)

Reply:
ok